### ISSUE TITLE
`@chpwd_dir-history-var` can't rename ... to `$HISTFILE`

### ISSUE THREAD

**Issue by lmburns** [NICHOLAS85/dotfiles#1]
Occasionally I will get this error when switching directories. I was wondering if you got it, and if so, how to fix it?

```sh
@chwpd_dir-history-var:19: can't rename /Users/lucasburns/.config/zsh/zinit/polaris/share/per-directory-history/Users/lucasburns/.config/zsh/zinit/plugins/Property404---lax/zsh-per-directory-history.new to $HISTFILE
```

Also, I noticed that I will get this error if I try and run `zinit recently 1 day`.
I fixed the `zinit` issue by adding:

```zsh
unfunction @append_dir-history-var
unfunction @chwpd_dir-history-var
```

I tried something like at the beginning of the function, but neither worked.
```zsh
[[ ${chpwd_functions[(re)_per-directory-history-change-directory]} ]] &&
    chpwd_functions=( ${chpwd_functions[@]:#_per-directory-history-change-directory} )
[[ ${preexec_functions[(re)_per-directory-history-preexec]} ]] &&
    preexec_functions=( ${preexec_functions[@]:#_per-directory-history-preexec})

# or
add-zsh-hook -d chpwd _per-directory-history-change-directory
add-zsh-hook -d preexec _per-directory-history-preexec
```

And then they could be added back at the end.

**Comment by NICHOLAS85** [comment]
Interesting, I haven't run into that issue myself but I'm happy to try to help you resolve it. First, since we are working with your HISTFILE which can be valuable to some, I recommend backing it up just in case its contents get deleted while trying to debug your issue. It most likely won't but it doesn't hurt to have a backup.

Just a couple of questions to start out with:
Do you use the per-directory history functionality? In my config, it is used to provide directory relevant autosuggestions as well as be switched between with global history after pressing a keybind. If you don't want to use either of these features then I can help de-integrate the plugin from your config.
What are the contents of your zshrc and related files, Are you trying to merge a personal config with the config located in this repo or just testing this repos config on its own?
Is the error message the same every time, as in the same `/Users/lucasburns/.../zsh-per-directory-history.new` filepath shown each time?

**Comment by lmburns** [comment]
Yes I use the per directory history functionality. I've actually been using some aspects of your configuration for quite some time now, and I just recently started to use `zinit` for more than just the simple downloading plugins. I wanted to see what other functions it was capable of and tried using the `zi recently` call and it won't complete because of the error. So that's why I added the `unfunction` mentioned above to `zinit`

I'm on my phone now or I would type out the function names fully, but I use both of the functions you have that start with `@`. The one mentioned in the above comment in the error is the function that is causing the issue. I also use the same patch you applied to that same repo where the per directory history is found. 

The error with just randomly show up it seems like. It doesn't happen every time I `cd` somewhere, it's just occasionally. I think I've noticed that it is directories that I do not visit as often or it seems to happen more frequently if I `jump` to a directory using `wfxr/formarks` plugin. Which is just a bookmark file that is popped up with `fzf` and then change to the directory when enter is pressed. 

If there is an error, it is the same error yes, just with a different directory displayed. Another thing I've noticed (may not be related) is that sometimes when switching directories and then going to type a command in the new directory, there is substantial lag in the `chpwd_ls` call to display the directories. 

One thing that will guarantee the error is to use `zinit recently 1 day` or any time frame. It may get a couple to show up, but it will 100% display that same error and then stop checking other directories. I'll be home in around an hour and can comment all of the relevant functions and what they do. 

I am on a Mac if that makes any difference at all. Now that I'm typing all of this out I'm wondering if the `fc` command has different flag meanings on Linux and Mac. [Here](https://github.com/lmburns/dotfiles/blob/master/.config/zsh/.zshrc) is my `zshrc` for the moment. There is still a push I need to make but this configuration will still produce the error.

**Comment by lmburns** [comment]
I thought that it could have somethign to do with `OMZ::lib/history.zsh` or `MichaelAquilina/zsh-history-filter`. However, I deleted both of them and the error still shows up when calling `zi recently 1 day`. And the error goes away whenever `unfunction @chwpd_dir-history-var` is placed within `.zinit-recently` function.

I am unsure what the digit after the function's name is referring (`@chwpd_dir-history-var:22:`
) to because the file is only a couple lines long.

Also, I just got another error and it starts with `zsh` instead of `@chwpd_dir-history-var`.
```zsh
zsh: can't rename /Users/lucasburns/.config/zsh/zinit/polaris/share/per-directory-history/Users/lucasburns/Documents/cv/Awesome-CV/zsh-per-directory-history.new to $HISTFILE
```
I've also noticed that it seems to happen more frequently when using `autojump` or `zoxide`, as well as the `wfxr/formarks` jumping to directories plugin.

I know you are aware of these below that are being referenced, because you wrote them. But here is what I think has to do with the problem. When commenting out every other plugin in my `zshrc` the error still seemed to be there. I also looked up to see if either `-a` or `-p` were any different for `fc` on Linux and macOS and they're not. This post was more so just to see if you ever got the error.

`@chwpd_dir-history-var:22:`
```zsh
emulate -L zsh

if [[ -r "$_per_directory_history_path" ]]; then
  fc -a -p "$_per_directory_history_path"
fi

dir_history=("${history[@]}")
```

`@append_dir-history-var`
```zsh
emulate -L zsh

dir_history=("${1%%$'\n'}" "${(u)dir_history[@]}" )
```

`zshrc`
```zsh
if ! [[ $MYPROMPT = dolphin ]]; then
  autoload -Uz chpwd_recent_dirs add-zsh-hook
  add-zsh-hook chpwd chpwd_recent_dirs
  zstyle ':chpwd:*' recent-dirs-file "${TMPDIR}/chpwd-recent-dirs"
  dirstack=($(awk -F"'" '{print $2}' ${$(zstyle -L ':chpwd:*' recent-dirs-file)[4]} 2>/dev/null))
  [[ ${PWD} = ${HOME}  || ${PWD} = "." ]] && (){
    local dir
    for dir ($dirstack){
      [[ -d "${dir}" ]] && { cd -q "${dir}"; break }
    }
  } 2>/dev/null
fi

# possibly referenced
[[ $MYPROMPT != dolphin ]] && add-zsh-hook chpwd chpwd_ls
```

And the plugin itself:
```zsh
  atload'add-zsh-hook chpwd @chwpd_dir-history-var;
  add-zsh-hook zshaddhistory @append_dir-history-var; @chwpd_dir-history-var' \
    kadaan/per-directory-history \
```
[per_directory_history patch](https://github.com/lmburns/dotfiles/blob/master/.config/zsh/patches/per-directory-history.patch)

**Comment by NICHOLAS85** [comment]
From what it looks like the `fc -a -p` line is causing an error on your system but I am unsure why at the moment. For now, I've pushed a change to `@chwpd_dir-history-var` which exits the function if it detects its running within another function. This is a temporary fix as I actually do want `@chwpd_dir-history-var` to run in certain functions but for now, it can restore `zinit recently` functionality for you.
https://github.com/NICHOLAS85/dotfiles/commit/28deaa53ad1a987e8f1e3880c4b5a581b9ccc628#diff-801d6c898b532185e8872f461bfb2ccf11ad8bbdd550d5496bf657f529bd1decR1-R7

**Comment by lmburns** [comment]
It did fix the error of not being able to rename the `$HISTFILE`, however there is another error now whenever running `zinit recently`. By the way, I wish I knew `zsh` as well as you. I had no idea about `${funcstack[@]}` or `${functrace[@]}`, etc. 

I have [`zoxide`](https://github.com/ajeetdsouza/zoxide) as well, and the error now has to do with that. It is:
```zsh
$ zi recently '1 day'                                                           【 ~/.conf/zs/zi/bin】─╯
__zoxide_pwd:1: write error: bad file descriptor
error: The argument '<path>' requires a value but none was supplied

USAGE:
    zoxide add <path>

For more information try --help
@zinit-scheduler:72: write error: bad file descriptor
__zoxide_pwd:1: write error: bad file descriptor
error: The argument '<path>' requires a value but none was supplied

USAGE:
    zoxide add <path>

For more information try --help
chpwd_recent_filehandler:29: cannot duplicate fd 1: too many open files
_flush_ysu_buffer:3: cannot duplicate fd 1: too many open files
zsh-notify-after-command:9: 1: bad file descriptor
```
It seems to be calling `zoxide` somewhere, and I am unsure about the file descriptor stuff. Here are the currently loaded hooks I have running:
```zsh
$ zhooks                                                                    【25s  ~/.conf/zs/zi/bin】─╯
chpwd_functions:
chpwd_recent_dirs
chpwd_ls
autojump_chpwd
@zinit-scheduler
_per-directory-history-change-directory
@chwpd_dir-history-var
_autoenv_chpwd_handler
__zoxide_hook

precmd_functions:
_p9k_do_nothing
iterm2_precmd
_flush_ysu_buffer
zsh-notify-after-command
_p9k_precmd

preexec_functions:
_p9k_preexec1
iterm2_preexec
_check_aliases
_check_global_aliases
_check_git_aliases
_per-directory-history-preexec
_zsh_highlight_preexec_hook
zsh-notify-before-command
_p9k_preexec2

zshaddhistory_functions:
_per-directory-history-addhistory
@append_dir-history-var

zshaddhistory () {
        emulate -L zsh
        whence ${${(z)1}[1]} >| /dev/null || return 1
        [[ ${1%%$'\n'} != ${~HISTORY_IGNORE} ]]
}

zshexit_functions:
_gitstatus_cleanup_POWERLEVEL9K-_p9k_
```
I think that seems to be quite a bit of functions. Would the best way to debug this be to use `autoenv` and create a directory with the same `zshrc` and slowly add things? Also, is it possible to use something like `tail -f` on the currently running functions for `$funcstack[@]`?

**Comment by NICHOLAS85** [comment]
Haha honestly all my zsh knowledge comes from googling and trial and error, the internet is a great resource.
Quick question, What version of Zsh are you running? Slowly recreating your config from a brand new zshrc will definitely help in narrowing down what exactly is causing the issue. You could try running `zsh -x 2>$TMPDIR/zshlog` and then in a different window `tail -f $TMPDIR/zshlog` to view what is being run during a live session. It can get quite verbose though.

**Comment by lmburns** [comment]
I'm running `5.8`. I've got a similar configuration on my server that runs Debian and I have never had an issue with the history file doing this.  And yes that is quite verbose. I will try and recreate the configuration slowly. I tried this the other day and it was doing it with only the `per-directory-history` plugin enabled. 

Also with the googling of zsh stuff, do/did you use the sourceforge resource very much? Because whenever I try and google some stuff specifically to `zsh`, it seems more difficult to find answers, specifically on stackoverflow. I wrote [this little program](https://github.com/lmburns/dotfiles/blob/master/mybin/zm) to try and help parse the man page because I need to look up flags to use them because I've only got maybe 1/8th memorized.

**Comment by NICHOLAS85** [comment]
On the googling zsh stuff, I mainly use [this refcard](http://www.bash2zsh.com/zsh_refcard/refcard.pdf) to find information on variable expansion or flags among the plethora of other things it quickly provides. I find reading the sources of various zsh projects helps me better understand how to go about doing certain things. And when googling, I end up rephrasing my query or messing with Google's search filtering until I find a StackOverflow or manpage which contains an answer I need. I haven't used Sourceforge myself and can't comment on its usefulness.

Have you had any luck finding what's going on with your zsh config?

**Comment by lmburns** [comment]
After your latest update on changing the `@chpwd...` function, I haven't had the error. I never did figure it out by eliminating everything from my `.zshrc` except the functions that were causing the issue because with only those enabled, the problem was still there. I'm glad you figured this out, so thank you. And I'm unsure as to why the issue is present on macOS, but not my server (Debian), and your linux distribution.

Edit: However, I should note that I know longer have errors stating that the `$HISTFILE` could not be popped or whatever it was. I do still have the `zoxide` error whenever `zi recently...` is called. Afterwards, trying to use something like `neovim` causes a segmentation fault. So I'm unsure there because  But if the source code is peeked at, `.zinit-recently() {...` only contains `pushd` and `popd`. The only other function it calls is `.zinit-any-colorify-as-uspl2`, which calls `.zinit-any-to-pid`, which calls `.zinit-util-shands-path`. I'll place these functions into a script and see if I can figure anything out here.

When putting these into a script and running the same thing, there is no error called. [Here it is](https://github.com/lmburns/dotfiles/blob/master/mybin/zirecently)
 
Final edit:
I could've sworn I did this before, I guess not, but all I did was add `-q` to the `popd` and `pushd` in `zi recently`. What you did for the `@chwpd...` functions fixed the error that said that the `$HISTFILE` couldn't be popped when changing directories. Issue resolved.
